### ISSUE TITLE
Recursive processes being created in update-port.sh

### ISSUE THREAD

**Issue by taylorcox75** [haugene/docker-transmission-openvpn#1479]
PIA, container gets messy after a few days (this is after 5). 1 maybe 2 torrents added in total.  Torrenting still works and the port was still open, but there are extra processes being spawned somewhere in the code. 

`ps aux`

```
    1 root         0:00 dumb-init /etc/openvpn/start.sh
    6 root         0:58 openvpn --script-security 2 --up-delay --up 
    219 root       0:00 {update-port.sh} /bin/bash /etc/openvpn/pia/update-port.sh
    598 root       0:00 /bin/bash
    3732 root      0:00 {update-port.sh} /bin/bash /etc/openvpn/pia/update-port.sh
    5409 root      0:00 {update-port.sh} /bin/bash /etc/openvpn/pia/update-port.sh
    5698 root      0:00 {update-port.sh} /bin/bash  /etc/openvpn/pia/update-port.sh
    5708 abc       0:05 /usr/bin/transmission-daemon -g /data/transmissiom-home --logfile /data/transmission-home/transmission.log
    6025 root      0:00 sleep 1800
    6058 root      0:00 sleep 1800
    6146 root      0:00 sleep 1800
    6206 root      0:00 sleep 1800
    6301 root      0:00 /bin/bash
    6306 root      0:00 ps
```

Container as mentioned is still connected and port is still open (everything works in other words)

```
bash-5.0# curl ipinfo.io                                                          {
  "ip": "195.246.120.131",
  "hostname": "195-246-120-131-static.serverhotell.net",
  "city": "Stockholm",
  "region": "Stockholm",
  "country": "SE",
  "loc": "59.3294,18.0687",
  "org": "AS42708 GleSYS AB",
  "postal": "100 05",
  "timezone": "Europe/Stockholm",
  "readme": "https://ipinfo.io/missingauth"
```

Dockerfile is pretty much the default with some custom paths. 

Run via docker-compose. 
`docker-compose up`

Then when it looks pretty, exit and 

`docker-compose start`

Or running detached

`docker-compose up -d`

Pretty much identical to the Readme example with the following changes

```
cap_add:
    - NET_ADMIN
  ports:
    - '9091:9091'
  dns:
    - '8.8.8.8'
    - '8.8.4.4'
 volumes:
   - /etc/localtime:/etc/localtime:ro
environment:
- OPENVPN_CONFIG=Sweden
- OPENVPN_OPT=--inactive 3600 --ping 10 --ping-exit 60
```

Running headless debian buster and issue started occuring right after the updated pia scripts were pushed maybe a few weeks ago. 

Have done several "delete and purge everything and build from scratch no cache" builds and issue persists. 

edit: Runs 24/7. No reboots of system or container / network outages

**Comment by haugene** [comment]
Thanks for reporting this. For now I'm prioritizing helping out where people report on stuff not working. Need to get an overview and fix whatever was broken when 3.0 was released. But will come back to this, and I agree it should be cleaned up :+1:

**Comment by haugene** [comment]
:point_up: Seems someone beat me to it. And there is nothing better :smiley: 
Can you test the changes using the dev tag @taylorcox75 and confirm that it's fixed? Then I'll merge to master, just need to confirm some other changes as well.

**Comment by mochman** [comment]
I've got the same issue, but the getSignature fix didn't work for me.  I think I'm seeing multiple update-port.sh processes because of #1447.  Sometimes I get the event_wait : Interrupted system call (code=4) and then healthcheck.sh kills the openvpn process and restarts it.  The issue is that update-port.sh doesn't get killed and when the new openvpn process comes up, it starts a new update-port.sh process.

**Comment by haugene** [comment]
Ah, then this might be because of the health check script as well. I wasn't aware that it actually signals openvpn to exit in an effort to get things back up and running again. It is being addressed in #1459 but for now you can just disable the healthcheck.

To disable health checks add `--no-healthcheck` to your [Docker run](https://docs.docker.com/engine/reference/run/#healthcheck) command or [in docker-compose](https://docs.docker.com/compose/compose-file/#healthcheck):
```
healthcheck:
  disable: true
```

**Comment by taylorcox75** [comment]
@haugene @mochman Can confirm. 

So good news and bad news. Bad news is the fix mentioned above didn't help this issue. 

Good news is I think I found the issue (Tl-DR at bottom)

Ports are good for 2 months. So we check every 30 minutes if the port is still open. We send a curl and get a reply.  PIA does every 15 minutes (they say this is the inactivity timeframe where you could get booted off the port). Few minor differences between ours. After every sleep we do the same basic thing, ours is a little outdated because of the whole next gen transition mess.  

Since tokens expire every 24 hours, our port ownership is tied to the payload and signature we got when we bound to the port. Now the official port script actually  has the user send over a .pem file for verification (and the server sends a certificate in the payload). We use an --unsecure flag which could be why some people are having curl errors when trying to bind a port


**But check this out** (from our update-ports.sh)
```
echo "token expiry $pf_token_expiry"                                              
pf_remaining=$((  $pf_token_expiry - $(date +%s) ))

Output:
token expiry 1610289885 
remaining = 5425148
```

English:
Token expiration: In 24 hours
Remaining: 62 days

The problem:

```
while true; do
  pf_remaining=$((  $pf_token_expiry - $(date +%s) ))

```

We're using the token which expires in 24 hours to determine if it's been 60 days or not. This while loop immediately goes and requests a new payload/signature every day which is why after a week I had five to six duplicate processes.


 Since I was still active on my original port, that's why it never was revoked. If somebody wasn't paying attention, after a couple weeks a few times a minute they would be trying to curl the server. 

Other side notes, sometimes a bad curl  (security warnings) will give you this

`remaining = -1604862953`

Which will also cause a duplicate process

I'm running a private fork now where I basically updated the port script to do what the pia repo does. Every 15 minutes

```
while true; do
  bind_port_response="$(curl -Gs -m 5 \
    --connect-to "PF_HOSTNAME::PF_GATEWAY:" \
    --cacert "ca.rsa.4096.crt" \
    --data-urlencode "payload=${payload}" \
    --data-urlencode "signature=${signature}" \
    "https://PF_HOSTNAME:19999/bindPort")"
    echo "$bind_port_response"

    export bind_port_response
    if [ "$(echo "$bind_port_response" | jq -r '.status')" != "OK" ]; then
      echo "The API did not return OK when trying to bind port. Exiting."
      exit 1
    fi
    echo Port $port refreshed on $(date). \
      This port will expire on $(date --date="$expires_at")

    # sleep 15 minutes
    sleep 900
done
 
``
In the .opvn all the Next Gen servers list the PF_HOSTAME the 4th row of the file

 ad.privacy.network, sweden.privacy.network, etc. 

TL-DR; Someone accidentally used the epoch time that expires every 24 hours, requesting a new port daily.  The other epoch time has a date  not for another 60 days.

**Comment by taylorcox75** [comment]
So a super TL-DR 

Line 157 goes to:
` pf_remaining=$((  $pf_remaining - $(date +%s) ))`

**Comment by RickyXwang** [comment]
@taylorcox75 Good find. I thought it was wrong when I look at the code, but I don't have PIA to test/play with it. If it's working for you and is pretty stable, can you do a PR?

**Comment by taylorcox75** [comment]
So this issue can be closed due to the health_check causing the actual problem with recursive processes. 

I did a pretty decent update to update_ports.sh earlier this week and will probably make a new enhancement issue to keep this one focused on the health check issue.

**Comment by haugene** [comment]
Am I reading this correctly if it sounds like you'll have a PR for the update-port script for PIA sometime soon?
If so - that would be great! I've been postponing doing it myself but there are a lot of issues there these days.

**Comment by taylorcox75** [comment]
Yea mostly code cleanup and streamlining/removing duplicates. I'm running about 150 GB through transmission in the next few days, if it holds up I'll create the request. I got sidetracked with work and was left running a spaghetti code version that had its own issues. 

 If it continues to hold its ground, I'll throw it in mid-week


Also opened #1592 for the side note that this issue got derailed on about the official next gen configuration scripts. 

Short version is I have a working version of it, but the official way picks the best server based on parameters the user gives it opposed to having the user pick their own server. Which could potentially be a turn off
